renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp

```diff
@@ -4720,6 +4720,109 @@ void WrappedVulkan::vkCmdEndQueryIndexedEXT(VkCommandBuffer commandBuffer, VkQue
   }
 }
 
+void WrappedVulkan::vkCmdBeginConditionalRenderingEXT(
+    VkCommandBuffer commandBuffer,
+    const VkConditionalRenderingBeginInfoEXT *pConditionalRenderingBegin)
+{
+  SCOPED_DBG_SINK();
+
+  VkConditionalRenderingBeginInfoEXT unwrappedConditionalRenderingBegin = *pConditionalRenderingBegin;
+  unwrappedConditionalRenderingBegin.buffer = Unwrap(unwrappedConditionalRenderingBegin.buffer);
+
+  SERIALISE_TIME_CALL(ObjDisp(commandBuffer)
+                          ->CmdBeginConditionalRenderingEXT(Unwrap(commandBuffer),
+                                                            &unwrappedConditionalRenderingBegin));
+
+  if(IsCaptureMode(m_State))
+  {
+    VkResourceRecord *record = GetRecord(commandBuffer);
+
+    CACHE_THREAD_SERIALISER();
+
+    SCOPED_SERIALISE_CHUNK(VulkanChunk::vkCmdBeginConditionalRenderingEXT);
+    Serialise_vkCmdBeginConditionalRenderingEXT(ser, commandBuffer, pConditionalRenderingBegin);
+
+    record->AddChunk(scope.Get());
+
+    VkResourceRecord *buf = GetRecord(pConditionalRenderingBegin->buffer);
+
+    record->MarkResourceFrameReferenced(buf->GetResourceID(), eFrameRef_Read);
+    record->MarkResourceFrameReferenced(buf->baseResource, eFrameRef_Read);
+
+    if(IsActiveCapturing(m_State))
+      GetResourceManager()->MarkResourceFrameReferenced(buf->GetResourceID(), eFrameRef_Read);
+  }
+}
+
+template <typename SerialiserType>
+bool WrappedVulkan::Serialise_vkCmdBeginConditionalRenderingEXT(
+    SerialiserType &ser, VkCommandBuffer commandBuffer,
+    const VkConditionalRenderingBeginInfoEXT *pConditionalRenderingBegin)
+{
+  SERIALISE_ELEMENT(commandBuffer);
+  SERIALISE_ELEMENT_LOCAL(BeginInfo, *pConditionalRenderingBegin);
+
+  Serialise_DebugMessages(ser);
+
+  SERIALISE_CHECK_READ_ERRORS();
+
+  if(IsReplayingAndReading())
+  {
+    m_LastCmdBufferID = GetResourceManager()->GetOriginalID(GetResID(commandBuffer));
+
+    if(IsActiveReplaying(m_State) && InRerecordRange(m_LastCmdBufferID))
+    {
+      m_RenderState.conditionalRendering.buf = GetResID(BeginInfo.buffer);
```

Comment:
This is a tricky one. Normally for state setting you'd need an extra call to `ShouldUpdateRenderState` to check if the current renderstate should be updated. Otherwise if your state is set in a secondary command buffer, it will be trampled by any later calls in the primary command buffer while we're re-recording and tracking the state. Current renderpass/subpass/framebuffer is special because that can't be set in the secondary buffer so it's set by the primary.

In this case the conditional rendering can either be set independently in the secondary command buffer _or_ inherited, which I don't think happens with anything else (either state is entirely inherited or entirely reset).

I _think_ the solution would be to do this:
1. This `m_RenderState` is protected by `if(ShouldUpdateRenderState(m_LastCmdBufferID))` which will handle the simple cases correctly - everything but inheritance.
2. Track conditional rendering state in the `m_BakedCmdBufferInfo[m_LastCmdBufferID].state` struct which tracks a subset of the full state per-command buffer instead of globally. Similar to how renderpass/framebuffer is currently.
3. When it comes to `vkCmdExecuteCommands`, if one of the command buffers is one we're partially executing **and** is inheriting conditional rendering state, then copy the `m_BakedCmdBufferInfo[m_LastCmdBufferID].state` conditional state over the `m_RenderState` conditional state. The spec says conditional rendering must be started and stopped in the same command buffer, so we don't have to worry about the state changing within a secondary if it's inherited.

I know that's pretty in-depth :smile:. If you want, in this PR you can just do part 1 and I'll do the rest in a follow-on commit.

Reply:
Ughhhh, yes, I have missed this. And it reminded me about what I have missed when I wrote Vulkan CTS tests for this extension...

Will try to implement it the right way.

Reply:
I implemented inherited conditional rendering the way you suggested - it works!